kong/runloop/handler.lua

```diff
@@ -435,13 +457,44 @@ return {
         log(ngx.CRIT, "could not ensure API router is up to date: ", err)
 
       elseif api_router_version ~= version then
-        log(DEBUG, "rebuilding API router")
 
-        local ok, err = build_api_router(singletons.dao, version)
+        -- wrap api_router rebuilds in a per-worker mutex (via ngx.semaphore)
+        -- this prevents dogpiling the database during rebuilds in
+        -- high-concurrency traffic patterns
+        -- requests that arrive on this process during a api_router rebuild will be
+        -- queued. once the semaphore resource is acquired we re-check the
+        -- api_router version again to prevent unnecessary subsequent rebuilds
+        --
+        -- note the ten-second wait() timeout. we should be able to safely assume
+        -- that a api_router rebuild event should never take this long
+
+        local ok, err = build_api_router_semaphore:wait(10)
         if not ok then
-          api_router_err = err
-          log(ngx.CRIT, "could not rebuild API router: ", err)
+          return responses.send_HTTP_INTERNAL_SERVER_ERROR(
```

Comment:
As I was going through one last go over this, I started wondering if the best course of action in this scenario would be to log a warning/error, and fallthrough to the cache checking/building. We would still prevent stampede effects to the database in almost all cases, but if a situation ever occurs where a particular rebuild gets slowed down, we don't prevent another request to rebuild the router on its own - on Cassandra, a subsequent router rebuild would hit another peer, for instance. And a client with a large enough timeout would avoid having to deal with a 500. At the very least, I think we should not send this error message to the client (as done by `send_HTTP_INTERNAL_SERVER_ERROR(err)`), but just logging it for the administrator to see.

Reply:
> This is already the case; send_HTTP_INTERNAL_SERVER_ERROR does not write the err message to the client:

Oh yes, that's great then! I should know, I wrote this module... -_-

> I'm not sure I see the value in retrying the router build following a mutex acquisition failure

The point that I am trying to convey here is that we should follow a "best effort" practice with regards to the proxy's resiliency - at all times. I feel like hard-coding the semaphore timeout at 10 seconds is very arbitrary, and combined with the hard failure approach taken in case of a timeout, becomes a recipe that is bound to at some point provoke 500 errors for clients, when we could maybe avoid it.
When confronted to this sort of problem (need for a retry, which could cause a cascading failure), I usually think of implementing the proper retry policy. Something much more precautionary with regards to retrying a failed operation (with the proper retry delays, ideally increasing exponentially). In this particular case (and Saturday morning), I failed to elaborate, and wanted to spark a discussion about the topic in general (making the best effort).

In this particular case, I think that granting one or a few more resources to the semaphore (instead of suddenly unlocking every waiting request) would allow for another request to attempt rebuilding the router, and maybe successfully so, thus preventing all waiting requests from resulting in a 500 just because the first try failed (or is lagging behind).
Granted, 10 seconds is a large timeout in the first place, but:

1. most clients would be willing to wait up to 30s, which would allow for the "retry" to work
2. the retry approach would allow us to reduce the initial timeout value, and trigger the "retry mode" faster

Again, a retry mode can be useful because it would connect to a different Cassandra peer than the first one (due to the load balancing policy), and it certainly wouldn't hurt to retry a query on postgres either, especially if some middleware is involved between Kong and the instance, as it can be the case. The "best-practise" approach (that we need to follow in the proxy, but even more so in the core, and even more so in the router component), combined with the above arguments, kinds of add up and made me raise this point.

Reply:
Talked IRL yesterday; my concern here is that introducing complex retry logic wrapped around sempahore resources introduces more possibility for buggy code- if we make a mistake and lose track of something, we could end up with very odd behavior (or lose the ability for a worker to acquire a resource at all). The solution we've potentially settled on is keeping the timeout + mutex logic, and using the Kong configuration's cassandra_timeout when Kong is talking to Cassandra (for Postgres, we can use a default value of 60s since pgmoon doesn't support setting socket timeouts).

Gonna sit on this today and think it through, then will update the PR accordingly.

Reply:
@bungle Oh nice! And... I just realized _I_ contributed this feature to pgmoon a couple years ago, and I totallty forgot about it and did not document it nor implemented the option in Kong :man_facepalming: :man_facepalming: :man_facepalming: Damn...

Reply:
Docs PR for `pgmoon:settimeout()`: https://github.com/leafo/pgmoon/pull/75

In the meantime, Kong does not have a `pg_timeout` config option, so we can still go ahead with the plan described by @p0pr0ck5 until we add support for it.

Thanks @bungle!

Reply:
@thibaultcha both `:wait` calls have been updated accordingly